diff --git a/Jamfile b/Jamfile
index 9c740467..bba7864c 100644
--- a/Jamfile
+++ b/Jamfile
@@ -110,7 +110,14 @@ rule make_libusrsctp_msvc ( targets * : sources * : properties * )
 {
 	local VARIANT = [ feature.get-values <variant> : $(properties) ] ;
 	VARIANT on $(targets) = $(VARIANT) ;
-	BUILD_DIR on $(targets) = "build-$(VARIANT)" ;
+
+	local ADDRESS_MODEL = [ feature.get-values <address-model> : $(properties) ] ;
+	local ARCH = x64 ;
+	if $(ADDRESS_MODEL) = 32
+	{ ARCH = Win32 ; }
+	ARCH on $(targets) = $(ARCH) ;
+
+	BUILD_DIR on $(targets) = "build-$(ARCH)-$(VARIANT)" ;
 }
 actions make_libusrsctp_msvc
 {
@@ -118,7 +125,7 @@ actions make_libusrsctp_msvc
     cd $(CWD)/deps/usrsctp
     mkdir $(BUILD_DIR)
     cd $(BUILD_DIR)
-    cmake -G "Visual Studio 17 2022" -Dsctp_werror=0 -Dsctp_build_shared_lib=0 -Dsctp_build_programs=0 -Dsctp_inet=0 -Dsctp_inet6=0 ..
+    cmake -G "Visual Studio 17 2022" -A $(ARCH) -Dsctp_werror=0 -Dsctp_build_shared_lib=0 -Dsctp_build_programs=0 -Dsctp_inet=0 -Dsctp_inet6=0 ..
     msbuild usrsctplib.sln /property:Configuration=$(VARIANT)
     cd %OLDD%
     cp $(CWD)/deps/usrsctp/$(BUILD_DIR)/usrsctplib/Release/usrsctp.lib $(<)
@@ -165,14 +172,21 @@ rule make_libjuice_msvc ( targets * : sources * : properties * )
 {
 	local VARIANT = [ feature.get-values <variant> : $(properties) ] ;
 	VARIANT on $(targets) = $(VARIANT) ;
+
+	local ADDRESS_MODEL = [ feature.get-values <address-model> : $(properties) ] ;
+	local ARCH = x64 ;
+	if $(ADDRESS_MODEL) = 32
+	{ ARCH = Win32 ; }
+	ARCH on $(targets) = $(ARCH) ;
+
 	if <gnutls>on in $(properties)
 	{
-		BUILD_DIR on $(targets) += "build-gnutls-$(VARIANT)" ;
+		BUILD_DIR on $(targets) += "build-gnutls-$(ARCH)-$(VARIANT)" ;
 		CMAKEOPTS on $(targets) = "-DUSE_NETTLE=1" ;
 	}
 	else
 	{
-		BUILD_DIR on $(targets) += "build-openssl-$(VARIANT)" ;
+		BUILD_DIR on $(targets) += "build-openssl-$(ARCH)-$(VARIANT)" ;
 		CMAKEOPTS on $(targets) = "-DUSE_NETTLE=0" ;
 	}
 }
@@ -182,7 +196,7 @@ actions make_libjuice_msvc
     cd $(CWD)/deps/libjuice
     mkdir $(BUILD_DIR)
     cd $(BUILD_DIR)
-    cmake -G "Visual Studio 17 2022" $(CMAKEOPTS) ..
+    cmake -G "Visual Studio 17 2022" -A $(ARCH) $(CMAKEOPTS) ..
     msbuild libjuice.sln /property:Configuration=$(VARIANT)
     cd %OLDD%
     cp $(CWD)/deps/libjuice/$(BUILD_DIR)/Release/juice-static.lib $(<)
